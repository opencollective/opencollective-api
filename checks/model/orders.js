import '../../server/env';

import { sequelize } from '../../server/models';

import { runCheckThenExit } from './_utils';

async function checkDuplicateNonRecurringContribution() {
  const message = 'Duplicate non-recurring Contribution (no auto fix)';

  const results = await sequelize.query(
    `SELECT COUNT(*), o."id"
     FROM "Transactions" t
     INNER JOIN "Orders" o ON o."id" = t."OrderId"
     WHERE t."deletedAt" IS NULL
     AND t."createdAt" > '2024-01-01'
     AND t."OrderId" IS NOT NULL
     AND t."kind" = 'CONTRIBUTION'
     AND t."type" = 'CREDIT'
     AND t."RefundTransactionId" IS NULL
     AND o."SubscriptionId" IS NULL
     GROUP BY o."id"
     HAVING COUNT(*) > 1`,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results.length > 0) {
    // Not fixable
    throw new Error(message);
  }
}

export async function checkOrders() {
  await checkDuplicateNonRecurringContribution();
}

if (!module.parent) {
  runCheckThenExit(checkOrders);
}
