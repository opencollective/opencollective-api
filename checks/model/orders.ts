import '../../server/env';

import { sequelize } from '../../server/models';

import { runAllChecksThenExit } from './_utils';

async function checkDuplicateNonRecurringContribution() {
  const message = 'Duplicate non-recurring Contribution (no auto fix)';

  const results = await sequelize.query(
    `SELECT COUNT(*), o."id"
     FROM "Transactions" t
     INNER JOIN "Orders" o ON o."id" = t."OrderId"
     WHERE t."deletedAt" IS NULL
     AND t."createdAt" > '2024-01-01'
     AND t."OrderId" IS NOT NULL
     AND t."kind" = 'CONTRIBUTION'
     AND t."type" = 'CREDIT'
     AND t."RefundTransactionId" IS NULL
     AND o."SubscriptionId" IS NULL
     GROUP BY o."id"
     HAVING COUNT(*) > 1`,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results.length > 0) {
    // Not fixable
    throw new Error(message);
  }
}

async function checkPaidOrdersWithNullProcessedAt({ fix = false } = {}) {
  const message = 'Paid Order with null processedAt';

  const results = await sequelize.query(
    `
    SELECT id, "updatedAt"
    FROM "Orders"
    WHERE status = 'PAID'
    AND "processedAt" IS NULL
    ORDER BY "createdAt" DESC
  `,
    { type: sequelize.QueryTypes.SELECT, raw: true },
  );

  if (results.length > 0) {
    if (!fix) {
      throw new Error(message);
    } else {
      await sequelize.query(`
        UPDATE "Orders"
        SET "processedAt" = "updatedAt"
        WHERE status = 'PAID'
        AND "processedAt" IS NULL
      `);
    }
  }
}

export const checks = [checkDuplicateNonRecurringContribution, checkPaidOrdersWithNullProcessedAt];

if (!module.parent) {
  runAllChecksThenExit(checks);
}
