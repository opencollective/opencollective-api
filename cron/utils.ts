import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
