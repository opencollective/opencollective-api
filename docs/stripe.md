# Stripe

## Platform account

By default, a platform account is configured (`config.stripe.secret`). We use the test environment of Open Collective Inc.

By convention, the Organization with slug "opencollective" is using the Platform account configuration.

### Webhooks

Start the Stripe tool in your terminal:

```
stripe login
```

Pick the Open Collective Inc. account. Then:

```
stripe listen --forward-to localhost:3060/webhooks/stripe
```

Add the "webhook signing secret" to your environment:

```
STRIPE_WEBHOOK_SIGNING_SECRET=whsec_XXX
```

## Host accounts

### For Credit Card payments

The Host Stripe account simply needs to be connected to the Platform using the web interface.

The Platform will be triggering everything from its account using "Stripe Connect".

### For Virtual Cards

At the moment, it's not possible to interact with Virtual Cards from the Platform account, we need to get the Host secret key for this.

Checklist for Fiscal Hosts (test mode):

1. Ask Stripe to [activate Issuing](https://dashboard.stripe.com/setup/issuing/activate) on your Fiscal Host account
2. [Top Up](https://dashboard.stripe.com/test/topups) the Issuing balance with a reasonable amount (e.g.: $1000).
3. Create a generic Card Holder and make sure it's the only one (Organization, name it like the Fiscal Host)
4. Create a new dedicated [Restricted Secret Key](https://dashboard.stripe.com/test/apikeys), select write for all Issuing features (Name: Restricted Issuing)
5. Add the Secret Key to the Fiscal Host (`npm run script ./scripts/update-connected-account-stripe-token`)

### Webhooks

Start the Stripe tool in your terminal:

```
stripe login
```

- To test Credit Card payments. Pick the Platform (Open Collective Inc.) account.
- To test Virtual Cards. Pick the Fiscal Host (e.g.: Open Source Collective) account.

Then:

```
stripe listen --forward-to localhost:3060/webhooks/stripe
```

Add the "webhook signing secret" to your environment:

_If switching between Platform and Fiscal Host, you'll have to change the "webhook signing secret" back and forth._

```
STRIPE_WEBHOOK_SIGNING_SECRET=whsec_XXX
```
