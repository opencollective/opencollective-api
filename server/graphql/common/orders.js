import { isNil } from 'lodash';

import status from '../../constants/order_status';
import { purgeCacheForCollective } from '../../lib/cache';
import logger from '../../lib/logger';
import * as libPayments from '../../lib/payments';
import models from '../../models';

export async function addFunds(order, remoteUser) {
  if (!remoteUser) {
    throw new Error('You need to be logged in to add fund to collective');
  }

  if (order.totalAmount < 0) {
    throw new Error('Total amount cannot be a negative value');
  }

  const { collective, fromCollective, host } = order;
  order.collective = collective;

  let user;

  // @deprecated Users are normally not created inline anymore
  if (order.user && order.user.email) {
    logger.warn('addFundsToCollective: Inline user creation should not be used anymore');
    user = await models.User.findByEmail(order.user.email);
    if (!user) {
      user = await models.User.createUserWithCollective({
        ...order.user,
        currency: collective.currency,
        CreatedByUserId: remoteUser ? remoteUser.id : null,
      });
    }
  } else if (remoteUser) {
    user = remoteUser;
  }

  if (fromCollective.hasBudget()) {
    // Make sure logged in user is admin of the source profile, unless it doesn't have a budget (user
    // or host organization without budget activated). It's not an ideal solution though, as spammy
    // hosts could still use this to pollute user's ledgers.
    const isAdminOfFromCollective = remoteUser.isRoot() || remoteUser.isAdmin(fromCollective.id);
    if (!isAdminOfFromCollective && fromCollective.HostCollectiveId !== host.id) {
      const fromCollectiveHostId = await fromCollective.getHostCollectiveId();
      if (!remoteUser.isAdmin(fromCollectiveHostId) && !host.data?.allowAddFundsFromAllAccounts) {
        throw new Error(
          "You don't have the permission to add funds from accounts you don't own or host. Please contact support@opencollective.com if you want to enable this.",
        );
      }
    }
  }

  if (order.tier && order.tier.CollectiveId !== order.collective.id) {
    throw new Error(`Tier #${order.tier.id} is not part of collective #${order.collective.id}`);
  }

  const orderData = {
    CreatedByUserId: remoteUser.id || user.id,
    FromCollectiveId: fromCollective.id,
    CollectiveId: collective.id,
    totalAmount: order.totalAmount,
    currency: collective.currency,
    description: order.description,
    status: status.NEW,
    TierId: order.tier?.id || null,
    data: {},
  };

  // Handle specific fees
  if (!isNil(order.hostFeePercent)) {
    orderData.data.hostFeePercent = order.hostFeePercent;
  }

  if (!isNil(order.memo)) {
    orderData.data.memo = order.memo;
  }

  if (!isNil(order.processedAt)) {
    orderData.processedAt = order.processedAt;
  }

  const orderCreated = await models.Order.create(orderData);

  const hostPaymentMethod = await host.getOrCreateHostPaymentMethod();
  await orderCreated.setPaymentMethod({ uuid: hostPaymentMethod.uuid });

  await libPayments.executeOrder(remoteUser || user, orderCreated, {
    invoiceTemplate: order.invoiceTemplate,
    isAddedFund: true,
  });

  // Invalidate Cloudflare cache for the collective pages
  purgeCacheForCollective(collective.slug);
  purgeCacheForCollective(fromCollective.slug);

  return models.Order.findByPk(orderCreated.id);
}
