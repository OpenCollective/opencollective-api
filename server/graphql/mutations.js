import models from '../models';
import paymentsLib from '../lib/payments';
import emailLib from '../lib/email';
import responseStatus from '../constants/response_status';

import {
  GraphQLNonNull,
  GraphQLString
} from 'graphql';

import {
  ResponseType,
  TierType
} from './types';

import {
  ResponseInputType,
  TierInputType
} from './inputTypes';

// import { hasRole } from '../middleware/security/auth';
// import {HOST, MEMBER} from '../constants/roles';

const mutations = {
  createTier: {
    type: TierType,
    args: {
      groupSlug: { type: new GraphQLNonNull(GraphQLString) },
      eventSlug: { type: new GraphQLNonNull(GraphQLString) },
      tier: { type: TierInputType }
    },
    resolve(_, args, req) {
      const tier = args.tier;
      console.log(">>> _", _, "args", args, "req", req)
      return models.Event.getBySlug(args.groupSlug, args.eventSlug)
      .then(() => models.Tier.create(tier))
    }
  },
  createResponse: {
    type: ResponseType,
    args: {
      response: {
        type: ResponseInputType
      }
    },
    resolve(_, args) {

      let tier, user, event, responseModel, isPaidTier;
      const response = args.response;
      response.user.email = response.user.email.toLowerCase();

      const recordInterested = () => {
        return models.Event.getBySlug(response.group.slug, response.event.slug)
        // find or create user
        .then(() => models.User.findOne({
          where: {
            $or: {
              email: response.user.email,
              paypalEmail: response.user.email
            }
          }
        }))
        .then(u => u || models.User.create(response.user))
        .tap(u => user = u)

        // create response
        .then(() => models.Response.create({
          UserId: user.id,
          GroupId: event.Group.id,
          EventId: event.id,
          confirmedAt: new Date(),
          status: response.status,
          description: response.description
        }));
      }

      const recordYes = () => {
        return models.Tier.findOne({
          where: {
            id: response.tier.id,
          },
          include: [{
            model: models.Event,
            where: {
              slug: response.event.slug
            },
            include: [{
              model: models.Group,
              where: {
                slug: response.group.slug
              }
            }]
          }]
        })
        .then(t => {
          if (!t) {
            throw new Error(`No tier found with tier id: ${response.tier.id} for event slug:${response.event.slug} in collective slug:${response.group.slug}`);
          }
          tier = t;
          event = t.Event;
          isPaidTier = tier.amount > 0;
        })

        // check for available quantity
        .then(() => tier.checkAvailableQuantity(response.quantity))
        .then(enoughQuantityAvailable => enoughQuantityAvailable ? 
              Promise.resolve() : Promise.reject(new Error(`No more tickets left for ${tier.name}`)))

        // make sure if it's paid tier, we have a card attached
        .then(() => isPaidTier && !(response.user.card && response.user.card.token) ? 
          Promise.reject(new Error(`This tier requires a payment method`)) : Promise.resolve())
        
        // find or create user
        .then(() => models.User.findOne({
          where: {
            $or: {
              email: response.user.email,
              paypalEmail: response.user.email
            }
          }
        }))
        .then(u => u || models.User.create(response.user))
        .tap(u => user = u)
        
        // create response
        .then(() => models.Response.create({
          UserId: user.id,
          GroupId: event.Group.id,
          EventId: event.id,
          TierId: tier.id,
          confirmedAt: isPaidTier ? null : new Date(),
          quantity: response.quantity || 0,
          status: response.status,
          description: response.description
        }))
        .tap(rm => responseModel = rm)

        // record payment, if needed
        .then(responseModel => {
          if (tier.amount > 0) {
            // also sends out email 
            return paymentsLib.createPayment({
              user,
              group: event.Group,
              response: responseModel,
              payment: {
                stripeToken: response.user.card.token,
                amount: tier.amount * responseModel.quantity,
                currency: tier.currency,
                description: `${event.name} - ${tier.name}`,
              }
            })
          } else {
            // only send out email if no payment
            return emailLib.send(
              'ticket.confirmed',
              user.email,
              { user: user.info,
                group: event.Group.info,
                response: responseModel.info,
                event: event.info,
                tier: tier && tier.info
              });
          }
        })
        .then(() => Promise.resolve(responseModel))
      }

      switch (response.status) {
        case responseStatus.INTERESTED:
          return recordInterested();

        case responseStatus.YES:
          return recordYes();
        
        default:
          throw new Error(`Unknown status ${response.status}`)
      }

    }
  }
}

export default mutations;