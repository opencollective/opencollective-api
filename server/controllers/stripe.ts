import config from 'config';
import { Request, Response } from 'express';
import { toNumber } from 'lodash';

import { idEncode } from '../graphql/v2/identifiers';
import errors from '../lib/errors';
import logger from '../lib/logger';
import { reportErrorToSentry } from '../lib/sentry';
import models from '../models';
import { confirmOrder } from '../paymentProviders/stripe/checkout';

export async function checkoutCallback(
  req: Request<any, any, { expenseIds: Array<string>; hostId: string }>,
  res: Response,
): Promise<void> {
  try {
    if (!req.query?.order) {
      throw new errors.BadRequest('Request missing Order id');
    }
    const orderId = toNumber(req.query.order);
    const order = await models.Order.findByPk(orderId, {
      include: [{ model: models.Collective, as: 'collective' }],
    });

    if (!order || !order.data?.session?.id) {
      throw new errors.NotFound('Could not find Order');
    }

    const collective = order.collective;
    const confirmedOrder = await confirmOrder(order);
    if (confirmedOrder) {
      const OrderId = idEncode(order.id, 'order');
      res.redirect(`${config.host.website}/${collective.slug}/donate/success?OrderId=${OrderId}`);
    } else {
      // TODO: redirect back to the contribution flow step
      res.redirect(`${config.host.website}/${collective.slug}`);
    }
  } catch (e) {
    logger.error('Error on processing Stripe Checkout callback', e);
    reportErrorToSentry(e);
    res
      .status(e.code || 500)
      .send(e.toString())
      .end();
  }
}
