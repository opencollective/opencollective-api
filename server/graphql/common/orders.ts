import { isNil } from 'lodash';
import { InferCreationAttributes } from 'sequelize';

import status from '../../constants/order-status';
import { purgeCacheForCollective } from '../../lib/cache';
import * as libPayments from '../../lib/payments';
import models, { AccountingCategory, Collective, Tier, User } from '../../models';
import { OrderModelInterface } from '../../models/Order';
import { ValidationFailed } from '../errors';
import { getOrderTaxInfoFromTaxInput } from '../v1/mutations/orders';
import { TaxInput } from '../v2/input/TaxInput';

type AddFundsInput = {
  totalAmount: number;
  collective: Collective;
  fromCollective: Collective;
  host: Collective;
  description: string;
  memo: string;
  processedAt: Date;
  hostFeePercent: number;
  tier: Tier;
  invoiceTemplate: string;
  tax: TaxInput;
  accountingCategory?: AccountingCategory;
};

/*
 * Throws if the accounting category is not allowed for this order/host
 */
export const checkCanUseAccountingCategoryForOrder = (
  accountingCategory: AccountingCategory | undefined | null,
  hostId: number,
): void => {
  if (!accountingCategory) {
    return;
  } else if (accountingCategory.CollectiveId !== hostId) {
    throw new ValidationFailed('This accounting category is not allowed for this host');
  } else if (accountingCategory.kind && !['ADDED_FUNDS', 'CONTRIBUTION'].includes(accountingCategory.kind)) {
    throw new ValidationFailed(`This accounting category is not allowed for contributions and added funds`);
  }
};

export async function addFunds(order: AddFundsInput, remoteUser: User) {
  if (!remoteUser) {
    throw new Error('You need to be logged in to add fund to collective');
  }

  if (order.totalAmount < 0) {
    throw new Error('Total amount cannot be a negative value');
  }

  const { collective, fromCollective, host } = order;

  if (fromCollective.hasBudget()) {
    // Make sure logged in user is admin of the source profile, unless it doesn't have a budget (user
    // or host organization without budget activated). It's not an ideal solution though, as spammy
    // hosts could still use this to pollute user's ledgers.
    const isAdminOfFromCollective = remoteUser.isRoot() || remoteUser.isAdmin(fromCollective.id);
    if (!isAdminOfFromCollective && fromCollective.HostCollectiveId !== host.id) {
      const fromCollectiveHostId = await fromCollective.getHostCollectiveId();
      if (!remoteUser.isAdmin(fromCollectiveHostId) && !host.data?.allowAddFundsFromAllAccounts) {
        throw new Error(
          "You don't have the permission to add funds from accounts you don't own or host. Please contact support@opencollective.com if you want to enable this.",
        );
      }
    }
  }

  if (order.tier && order.tier.CollectiveId !== order.collective.id) {
    throw new Error(`Tier #${order.tier.id} is not part of collective #${order.collective.id}`);
  } else if (order.accountingCategory) {
    checkCanUseAccountingCategoryForOrder(order.accountingCategory, host.id);
  }

  const orderData: Partial<InferCreationAttributes<OrderModelInterface>> = {
    CreatedByUserId: remoteUser.id,
    FromCollectiveId: fromCollective.id,
    CollectiveId: collective.id,
    totalAmount: order.totalAmount,
    currency: collective.currency,
    description: order.description,
    status: status.NEW,
    TierId: order.tier?.id || null,
    AccountingCategoryId: order.accountingCategory?.id || null,
    data: {
      hostFeePercent: order.hostFeePercent,
    },
  };

  if (!isNil(order.memo)) {
    orderData.data.memo = order.memo;
  }

  if (!isNil(order.processedAt)) {
    orderData['processedAt'] = order.processedAt;
  }

  if (order.tax?.rate) {
    orderData.taxAmount = Math.round(orderData.totalAmount - orderData.totalAmount / (1 + order.tax.rate));
    orderData.data.tax = getOrderTaxInfoFromTaxInput(order.tax, fromCollective, collective, host);
  }

  // Added Funds are not eligible to Platform Tips
  orderData.platformTipEligible = false;

  const orderCreated = await models.Order.create(orderData);

  const hostPaymentMethod = await host.getOrCreateHostPaymentMethod();
  await orderCreated.setPaymentMethod({ uuid: hostPaymentMethod.uuid });

  await libPayments.executeOrder(remoteUser, orderCreated, {
    invoiceTemplate: order.invoiceTemplate,
    isAddedFund: true,
  });

  // Invalidate Cloudflare cache for the collective pages
  purgeCacheForCollective(collective.slug);
  purgeCacheForCollective(fromCollective.slug);

  return models.Order.findByPk(orderCreated.id);
}
