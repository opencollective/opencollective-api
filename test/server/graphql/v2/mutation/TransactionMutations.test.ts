import { expect } from 'chai';
import gqlV2 from 'fake-tag';
import { createSandbox } from 'sinon';

import * as orders from '../../../../../server/graphql/v1/mutations/orders';
import emailLib from '../../../../../server/lib/email';
import * as payments from '../../../../../server/lib/payments';
import stripe from '../../../../../server/lib/stripe';
import models from '../../../../../server/models';
import stripeMocks from '../../../../mocks/stripe';
import { fakeCollective, fakeOrder, fakeUser } from '../../../../test-helpers/fake-data';
import { graphqlQueryV2 } from '../../../../utils';
import * as utils from '../../../../utils';

const STRIPE_TOKEN = 'tok_123456781234567812345678';

describe('server/graphql/v2/mutation/TransactionMutations', () => {
  let sandbox,
    collectiveAdminUser,
    hostAdminUser,
    randomUser,
    collective,
    order1,
    order2,
    transaction1,
    transaction2,
    sendEmailSpy,
    refundTransactionSpy;

  before(async () => {
    await utils.resetTestDB();
  });

  before(() => {
    sandbox = createSandbox();
    sandbox.stub(stripe.customers, 'create').callsFake(() => Promise.resolve({ id: 'cus_BM7mGwp1Ea8RtL' }));
    sandbox.stub(stripe.customers, 'retrieve').callsFake(() => Promise.resolve({ id: 'cus_BM7mGwp1Ea8RtL' }));
    sandbox.stub(stripe.tokens, 'create').callsFake(() => Promise.resolve({ id: 'tok_1AzPXGD8MNtzsDcgwaltZuvp' }));
    sandbox.stub(stripe.paymentIntents, 'create').callsFake(() =>
      Promise.resolve({
        id: 'pi_1F82vtBYycQg1OMfS2Rctiau',
        status: 'requires_confirmation',
      }),
    );
    sandbox.stub(stripe.paymentIntents, 'confirm').callsFake(() =>
      Promise.resolve({
        charges: { data: [{ id: 'ch_1AzPXHD8MNtzsDcgXpUhv4pm' }] },
        status: 'succeeded',
      }),
    );
    sandbox.stub(stripe.balanceTransactions, 'retrieve').callsFake(() => Promise.resolve(stripeMocks.balance));
    sandbox.stub(stripe.refunds, 'create').callsFake(() => Promise.resolve('foo'));
    sandbox.stub(stripe.charges, 'retrieve').callsFake(() => Promise.resolve('foo'));

    sandbox
      .stub(stripe.paymentMethods, 'create')
      .resolves({ id: 'pm_123456789012345678901234', type: 'card', card: { fingerprint: 'fingerprint' } });
    sandbox
      .stub(stripe.paymentMethods, 'attach')
      .resolves({ id: 'pm_123456789012345678901234', type: 'card', card: { fingerprint: 'fingerprint' } });

    sendEmailSpy = sandbox.spy(emailLib, 'send');
    refundTransactionSpy = sandbox.spy(orders, 'refundTransaction');
  });

  after(() => sandbox.restore());

  before(async () => {
    collectiveAdminUser = await fakeUser();
    hostAdminUser = await fakeUser();
    randomUser = await fakeUser();
    collective = await fakeCollective();
    await collective.addUserWithRole(collectiveAdminUser, 'ADMIN');
    await collective.host.addUserWithRole(hostAdminUser, 'ADMIN');
    await collectiveAdminUser.populateRoles();
    await hostAdminUser.populateRoles();
    order1 = await fakeOrder({
      CollectiveId: collective.id,
    });
    order1 = await order1.setPaymentMethod({ token: STRIPE_TOKEN });
    order2 = await fakeOrder({
      CollectiveId: collective.id,
    });
    order2 = await order2.setPaymentMethod({ token: STRIPE_TOKEN });
    await models.ConnectedAccount.create({
      service: 'stripe',
      token: 'abc',
      CollectiveId: collective.host.id,
    });
    await payments.executeOrder(randomUser, order1);
    transaction1 = await models.Transaction.findOne({
      where: {
        OrderId: order1.id,
        type: 'CREDIT',
      },
    });
    await payments.executeOrder(randomUser, order2);
    transaction2 = await models.Transaction.findOne({
      where: {
        OrderId: order2.id,
        type: 'CREDIT',
      },
    });
    await models.Member.create({
      CollectiveId: collective.id,
      MemberCollectiveId: randomUser.id,
      role: 'BACKER',
      CreatedByUserId: randomUser.id,
    });
  });

  afterEach(() => {
    refundTransactionSpy.resetHistory();
  });

  describe('refundTransaction', () => {
    const refundTransactionMutation = gqlV2/* GraphQL */ `
      mutation RefundTransaction($transaction: TransactionReferenceInput!) {
        refundTransaction(transaction: $transaction) {
          id
        }
      }
    `;

    it('refunds the transaction', async () => {
      const result = await graphqlQueryV2(
        refundTransactionMutation,
        {
          transaction: { legacyId: transaction1.id },
        },
        hostAdminUser,
      );

      result.errors && console.error(result.errors);
      expect(result.errors).to.not.exist;
      expect(result.data.refundTransaction.id).to.exist;
    });
  });

  describe('rejectTransaction', () => {
    const rejectTransactionMutation = gqlV2/* GraphQL */ `
      mutation RejectTransaction($transaction: TransactionReferenceInput!, $message: String) {
        rejectTransaction(transaction: $transaction, message: $message) {
          id
        }
      }
    `;

    it('should not refund the transaction if it has already been refunded but not rejected', async () => {
      await graphqlQueryV2(
        rejectTransactionMutation,
        {
          transaction: { legacyId: transaction1.id },
        },
        collectiveAdminUser,
      );

      expect(refundTransactionSpy.notCalled).to.be.true;
    });

    it('does not allow random user to reject', async () => {
      const result = await graphqlQueryV2(
        rejectTransactionMutation,
        {
          transaction: { legacyId: transaction2.id },
        },
        randomUser,
      );

      expect(result.errors).to.exist;
      expect(result.errors[0].message).to.match(/Cannot reject this transaction/);
    });

    it('rejects the transaction', async () => {
      const message = 'We do not want your contribution';
      const result = await graphqlQueryV2(
        rejectTransactionMutation,
        {
          transaction: { legacyId: transaction2.id },
          message,
        },
        hostAdminUser,
      );

      const updatedOrder = await models.Order.findOne({
        where: { id: order2.id },
      });

      const memberships = await models.Member.findOne({
        where: {
          MemberCollectiveId: transaction2.FromCollectiveId,
          CollectiveId: transaction2.CollectiveId,
          role: 'BACKER',
        },
      });

      await utils.waitForCondition(() => sendEmailSpy.calledWith('contribution.rejected'));

      expect(result.errors).to.not.exist;
      expect(result.data.rejectTransaction.id).to.exist;
      expect(sendEmailSpy.calledWith('contribution.rejected')).to.be.true;
      expect(updatedOrder.status).to.eq('REJECTED');
      expect(memberships).to.be.null;
    });
  });
});
