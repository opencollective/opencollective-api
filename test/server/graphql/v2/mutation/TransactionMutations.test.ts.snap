// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refundTransaction legacy tests Stripe Transaction - for hosts created after September 17th 2017 should be able to refund a stripe transaction with zero decimal currencies 1`] = `
"
| type   | kind                    | isRefund | To             | From           | amount  | paymentFee | platformFee | tax   | netAmountInCollectiveCurrency |
| ------ | ----------------------- | -------- | -------------- | -------------- | ------- | ---------- | ----------- | ----- | ----------------------------- |
| DEBIT  | HOST_FEE                | false    | Scouts d'Arlon | WWCode         | -50000  | 0          | 0           |       | -50000                        |
| CREDIT | HOST_FEE                | false    | WWCode         | Scouts d'Arlon | 50000   | 0          | 0           |       | 50000                         |
| DEBIT  | CONTRIBUTION            | false    | Phil Mod       | Scouts d'Arlon | -450000 | -17500     | -25000      | -7500 | -500000                       |
| CREDIT | CONTRIBUTION            | false    | Scouts d'Arlon | Phil Mod       | 500000  | -17500     | -25000      | -7500 | 450000                        |
| DEBIT  | PAYMENT_PROCESSOR_COVER | true     | WWCode         | Scouts d'Arlon | -17500  | 0          | 0           |       | -17500                        |
| CREDIT | PAYMENT_PROCESSOR_COVER | true     | Scouts d'Arlon | WWCode         | 17500   | 0          | 0           |       | 17500                         |
| DEBIT  | HOST_FEE                | true     | WWCode         | Scouts d'Arlon | -50000  | 0          | 0           | 0     | -50000                        |
| CREDIT | HOST_FEE                | true     | Scouts d'Arlon | WWCode         | 50000   | 0          | 0           | 0     | 50000                         |
| DEBIT  | CONTRIBUTION            | true     | Scouts d'Arlon | Phil Mod       | -500000 | 0          | 25000       | 7500  | -467500                       |
| CREDIT | CONTRIBUTION            | true     | Phil Mod       | Scouts d'Arlon | 467500  | 0          | 25000       | 7500  | 500000                        |"
`;

exports[`refundTransaction legacy tests Stripe Transaction - for hosts created after September 17th 2017 should create negative transactions without the stripe fee being refunded 1`] = `
"
| type   | kind                    | isRefund | To             | From           | amount  | paymentFee | platformFee | tax   | netAmountInCollectiveCurrency |
| ------ | ----------------------- | -------- | -------------- | -------------- | ------- | ---------- | ----------- | ----- | ----------------------------- |
| DEBIT  | HOST_FEE                | false    | Scouts d'Arlon | WWCode         | -50000  | 0          | 0           |       | -50000                        |
| CREDIT | HOST_FEE                | false    | WWCode         | Scouts d'Arlon | 50000   | 0          | 0           |       | 50000                         |
| DEBIT  | CONTRIBUTION            | false    | Phil Mod       | Scouts d'Arlon | -450000 | -17500     | -25000      | -7500 | -500000                       |
| CREDIT | CONTRIBUTION            | false    | Scouts d'Arlon | Phil Mod       | 500000  | -17500     | -25000      | -7500 | 450000                        |
| DEBIT  | PAYMENT_PROCESSOR_COVER | true     | WWCode         | Scouts d'Arlon | -17500  | 0          | 0           |       | -17500                        |
| CREDIT | PAYMENT_PROCESSOR_COVER | true     | Scouts d'Arlon | WWCode         | 17500   | 0          | 0           |       | 17500                         |
| DEBIT  | HOST_FEE                | true     | WWCode         | Scouts d'Arlon | -50000  | 0          | 0           | 0     | -50000                        |
| CREDIT | HOST_FEE                | true     | Scouts d'Arlon | WWCode         | 50000   | 0          | 0           | 0     | 50000                         |
| DEBIT  | CONTRIBUTION            | true     | Scouts d'Arlon | Phil Mod       | -500000 | 0          | 25000       | 7500  | -467500                       |
| CREDIT | CONTRIBUTION            | true     | Phil Mod       | Scouts d'Arlon | 467500  | 0          | 25000       | 7500  | 500000                        |"
`;

exports[`refundTransaction legacy tests Stripe Transaction - for hosts created before September 17th 2017 should create negative transactions with all the fees refunded 1`] = `
"
| type   | kind                    | isRefund | To             | From           | amount  | paymentFee | platformFee | tax   | netAmountInCollectiveCurrency |
| ------ | ----------------------- | -------- | -------------- | -------------- | ------- | ---------- | ----------- | ----- | ----------------------------- |
| DEBIT  | HOST_FEE                | false    | Scouts d'Arlon | WWCode         | -50000  | 0          | 0           |       | -50000                        |
| CREDIT | HOST_FEE                | false    | WWCode         | Scouts d'Arlon | 50000   | 0          | 0           |       | 50000                         |
| DEBIT  | CONTRIBUTION            | false    | Phil Mod       | Scouts d'Arlon | -450000 | -17500     | -25000      | -7500 | -500000                       |
| CREDIT | CONTRIBUTION            | false    | Scouts d'Arlon | Phil Mod       | 500000  | -17500     | -25000      | -7500 | 450000                        |
| DEBIT  | PAYMENT_PROCESSOR_COVER | true     | WWCode         | Scouts d'Arlon | -17500  | 0          | 0           |       | -17500                        |
| CREDIT | PAYMENT_PROCESSOR_COVER | true     | Scouts d'Arlon | WWCode         | 17500   | 0          | 0           |       | 17500                         |
| DEBIT  | HOST_FEE                | true     | WWCode         | Scouts d'Arlon | -50000  | 0          | 0           | 0     | -50000                        |
| CREDIT | HOST_FEE                | true     | Scouts d'Arlon | WWCode         | 50000   | 0          | 0           | 0     | 50000                         |
| DEBIT  | CONTRIBUTION            | true     | Scouts d'Arlon | Phil Mod       | -500000 | 0          | 25000       | 7500  | -467500                       |
| CREDIT | CONTRIBUTION            | true     | Phil Mod       | Scouts d'Arlon | 467500  | 0          | 25000       | 7500  | 500000                        |"
`;

exports[`refundTransaction legacy tests should create negative transactions with all the fees refunded 1`] = `
"
| type   | kind                    | isRefund | To             | From           | amount  | paymentFee | platformFee | tax   | netAmountInCollectiveCurrency |
| ------ | ----------------------- | -------- | -------------- | -------------- | ------- | ---------- | ----------- | ----- | ----------------------------- |
| DEBIT  | HOST_FEE                | false    | Scouts d'Arlon | WWCode         | -50000  | 0          | 0           |       | -50000                        |
| CREDIT | HOST_FEE                | false    | WWCode         | Scouts d'Arlon | 50000   | 0          | 0           |       | 50000                         |
| DEBIT  | CONTRIBUTION            | false    | Phil Mod       | Scouts d'Arlon | -450000 | -17500     | -25000      | -7500 | -500000                       |
| CREDIT | CONTRIBUTION            | false    | Scouts d'Arlon | Phil Mod       | 500000  | -17500     | -25000      | -7500 | 450000                        |
| DEBIT  | PAYMENT_PROCESSOR_COVER | true     | WWCode         | Scouts d'Arlon | -17500  | 0          | 0           |       | -17500                        |
| CREDIT | PAYMENT_PROCESSOR_COVER | true     | Scouts d'Arlon | WWCode         | 17500   | 0          | 0           |       | 17500                         |
| DEBIT  | HOST_FEE                | true     | WWCode         | Scouts d'Arlon | -50000  | 0          | 0           | 0     | -50000                        |
| CREDIT | HOST_FEE                | true     | Scouts d'Arlon | WWCode         | 50000   | 0          | 0           | 0     | 50000                         |
| DEBIT  | CONTRIBUTION            | true     | Scouts d'Arlon | Phil Mod       | -500000 | 0          | 25000       | 7500  | -467500                       |
| CREDIT | CONTRIBUTION            | true     | Phil Mod       | Scouts d'Arlon | 467500  | 0          | 25000       | 7500  | 500000                        |"
`;
